doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+   serve like pythons abstract base classes (ABC).
+
+These two rules are identical to the type choices made for example in the Julia
+language.
+It allows for the creation of a datatype hierarchy, but avoids issues with
+subclassing concrete DTypes directly:
+Logic such as ``can_cast`` does not cleanly inherit e.g. from a
+``Int64`` to a ``Datetime64`` even though the ``Datetime64`` could be seen
+as an integer with only a unit attached (and thus as a subclass).
+From the user perspective the introduction of abstract DTypes is
+thus rather the refusal of allowing subclassing
+and inheritance for concrete DTypes.
+While avoiding possible problems with subclassing and inheritance it still
+allows the transparent definition of a class/type hierarchy.
+
+As largely an implementation detail the DType class will require the homogeneous
+storage of additional, fixed, methods and information.
+This requires the creation of a ``DTypeMeta`` class.
+Each ``DType`` class is an instance of ``DTypeMeta`` with a well defined
+and extensible interface.
+
+.. image:: _static/dtype_hierarchy.svg
+
+
+Methods/Slots defined for each DType
+""""""""""""""""""""""""""""""""""""
+
+NEP 33 detailed that all logic should be defined through special methods
+on the DTypes.
+This section will list a specific set of methods (in the form of python methods).
+The C-side equivalent slot signature will be summarized below after proposing
+the general C-API for defining new Datatypes.
+Note that while the slots are defined as special python methods here, this is
+for convenience to the reader and *not* meant to imply the identical exposure
+as a Python API.
+This should be proposed as a separate NEP.
+
+
+Additional Information
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the below more detailed methods, the following general
+information is currently provided and will be defined on the class:
+
+* ``flexible``:
+  * Many dtypes are not flexible, i.e. they have a canonical
+    representation and casting from/to it is always safe.
+  * DTypes which are not flexible, must provide a canonical dtype instance
+    which should be a singleton.
+  * Flexible dtypes must define some additional methods.
+
+* ``is_native(self) -> Boolean`` method, and a corresponding attribute on the
+  dtype instance.
+  * Instead of byteorder, we may want an ``is_native`` flag (we could just
+    reuse the ISNBO flag – "is native byte order"),
+    this flag signals that the data is stored in the default/canonical way.
+    In practice this is always an NBO check, but generalization should be possible.
+    A use case is a complex-conjugated instance of Complex which is a
+    non-canonical representation, but may be native byte order.
+
+* ``ensure_native(self) -> dtype`` return a new dtype, or new reference to ``self``,
+  the returned dtype must fullfill ``dtype.is_native() is True`` and be otherwise
+  identical.
+
+* ``DType.type`` is the associated scalar type, also ``dtype.type`` is defined
```

Comment:
There is a pretty big philosophical and complicated issue around this, which on the one hand I would love to decide and discuss on the mailing list, but on the other hand feel it will be super hard to figure out.

Should scalars be encouraged/discouraged, and do we want array-scalars? For NumPy itself this does not matter, we will not change it. But in principle `dtype.type` could be `None` and in that case we would never return scalars (there would be no scalar after all!). On the other hand, I still somewhat like scalars, so the alternative is that I automatically create a simple scalar type for you if you set type to `None`...

The next thing is: For such new datatypes, we could try to make the ufunc `PyArray_Result` behaviour sane... I would love to just not decide it, but as soon as people create new datatypes there will be an implicit decision.